File: Data Structures & Algorithms/binary-tree-diameter/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        int[] res = {0};
        diameterOfBinaryTreeDfs(root, res);
        return res[0];
    }

    private int diameterOfBinaryTreeDfs(TreeNode node, int[] res) {
        if (node == null) return 0;
        int left = diameterOfBinaryTreeDfs(node.left, res);
        int right = diameterOfBinaryTreeDfs(node.right, res);

        res[0] = Math.max(res[0], left + right);

        return 1 + Math.max(left, right);
    }
}

File: Data Structures & Algorithms/trapping-rain-water/submission-1.js
class Solution {
    /**
     * @param {number[]} height
     * @return {number}
     */
    trap(height) {
        let l = 0 , r = height.length - 1;
        let lmax = height[l], rmax = height[r];
        let res = 0;
        while(l < r){
            if(lmax <= rmax){
                l++;
                lmax = Math.max(lmax,height[l]);
                res += lmax - height[l];
            }else{
                r--;
                rmax = Math.max(rmax,height[r]);
                res += rmax - height[r];
            }
        }
        return res;
    }
}
